### PR TITLE
Fancy cli

### LINKED ISSUE #7: Hyper-parameter optimisation for NB and SVM models
Find better values for the hyper-parameters of the NB and SVM models. The code should be part of https://github.com/msdslab/automated-systematic-review-simulations

### LINKED ISSUE #49: Add analysis class

### PR DESCRIPTION
**Pull request by qubixes**
Don't squash please!

Changelog:

- Introduce hooks for hyperopt optimization
-- Hyperopt is a package for hyperparameter optimization, and these hooks make it much easier to do the optimization of said hyperparameters.
-- The hyperopt package is optional.

- Model specific action are removed from the review package and moved in the model package.
- Models are now classes, instead of a generator function.

- Add HDF5 as a storage container for logging.
-- It works parallel to the still available JSON logger. It has an advantage in both speed and disk requirements (~x2.4).

- The JSON logger is updated, and it's version updated to 2.0, making it incompatible with log files created with an earlier version of ASReview. Automatically detected if this is the case.

- Add hooks for final_labels
-- This is a slightly obscure feature, where you can have for example two levels of inclusion:
--- After reading the abstract -> model trained on this
--- After reading full text
-- This is a simulation only feature that can be enabled by putting abstract_only=True in the [global_settings] section of a config file.

- Update the command line interface
-- Oracle only change.
-- This makes the command line interface much more useful, asking more questions and giving more options what to do next.
-- It uses the PyInquirer package, which unfortunately means that we cannot test oracle mode anymore with pytest (needs a TTY).

- The analysis part of the simulation project was ported over, since it can be generally useful.

- Documentation was updated.

Resolves #49 
Resolves #7

### CODE REVIEW

File: asreview/hdf5_logging.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+def _result_group(f, query_i):
+    try:
+        g = f[f'/results/{query_i}']
+    except KeyError:
+        g = f.create_group(f'/results/{query_i}')
+        g.attrs['creation_time'] = np.string_(datetime.now())
+    return g
+
+
+class HDF5_Logger(object):
```
Comment (J535D165) [inline asreview/hdf5_logging.py]: Naming conventions

File: asreview/hdf5_logging.py [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+
+    def __str__(self):
+        log_str = "Attributes:\n--------------\n"
+        for key, value in self.f.attrs.items():
+            log_str += f"{key}: {value}\n"
+        log_str += "\n\nGroups+datasets:\n---------------\n"
+        log_str += str(list(self.f.keys()))
+        return log_str
+
+    def todict(self):
```
Comment (J535D165) [inline asreview/hdf5_logging.py]: Inconsistent name

File: asreview/hdf5_logging.py [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+        ---------
+        fp: str
+            The file path to export the results to.
+
+        """
+        self.f['end_time'] = str(datetime.now())
+        self.f.flush()
+
+    def get(self, variable, query_i=None, idx=None):
+        "Get variables from hdf5 file."
```
Comment (J535D165) [inline asreview/hdf5_logging.py]: Docstring

File: asreview/hdf5_logging.py [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+        if variable == "train_idx":
+            array = np.array(g["train_idx"], dtype=np.int)
+        if array is None:
+            return None
+        if idx is not None:
+            return array[idx]
+        return array
+
+    def restore(self, fp, read_only=False):
+        "Restore logger from hdf5 file."
```
Comment (J535D165) [inline asreview/hdf5_logging.py]: Docstring

File: asreview/hdf5_logging.py [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+                    f"Log cannot be read: logger version {self.version}, "
+                    f"logfile version {log_version}.")
+            settings_dict = json.loads(self.f.attrs['settings'])
+            if "mode" in settings_dict:
+                self.settings = ASReviewSettings(**settings_dict)
+        except KeyError:
+            self.create_structure()
+
+    def create_structure(self):
+        "Initialize (empty) structure."
```
Comment (J535D165) [inline asreview/hdf5_logging.py]: docstring

File: asreview/logging.py [hunk 6]
```diff
@@ -1,230 +1,35 @@
...
-from datetime import datetime
-from pathlib import Path
+import os
 
-import asreview
-import numpy as np
-from asreview.settings import ASReviewSettings
 
+def get_logger_class(fp):
+    "Get logging class from file extension."
```
Comment (J535D165) [inline asreview/logging.py]: dosctring

File: asreview/json_logging.py [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+
+import numpy as np
+
+import asreview
+from asreview.settings import ASReviewSettings
+from asreview.analysis.statistics import _get_labeled_order
+from asreview.analysis.statistics import _get_last_proba_order
+
+
+class Dict_Logger(object):
```
Comment (J535D165) [inline asreview/json_logging.py]: Name. 

Should this be an abstract base class?

File: asreview/json_logging.py [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+
+    def close(self):
+        if not self.read_only:
+            self.save()
+
+    def save(self):
+        print(self)
+
+
+class JSON_Logger(Dict_Logger):
```
Comment (J535D165) [inline asreview/json_logging.py]: Name

File: asreview/json_logging.py [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+    def __init__(self, log_fp, *_, read_only=False, **__):
```
Comment (J535D165) [inline asreview/json_logging.py]: I don't like this syntax :) Does `def __init__(self, log_fp, read_only=False):` fail?

File: asreview/logging.py [hunk 10]
```diff
@@ -1,230 +1,35 @@
...
-
-        """
-        self._log_dict["settings"] = copy.deepcopy(vars(self.settings))
-        self._log_dict["settings"]["query_kwargs"].pop("pred_proba", None)
-        self._log_dict["settings"]["query_kwargs"].pop("query_src", None)
-        self._log_dict["settings"]["query_kwargs"].pop("current_queries", None)
-        self._log_dict.move_to_end("settings", last=False)
-        self._log_dict["time"]["end_time"] = str(datetime.now())
-        fp = Path(fp)
+    "Wrapper Logger class to easily get the right logger from file extension."
```
Comment (J535D165) [inline asreview/logging.py]: docstring

File: asreview/hdf5_logging.py [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+    def add_settings(self, settings):
+        self.settings = settings
+        self.f.attrs.pop('settings', None)
+        self.f.attrs['settings'] = np.string_(json.dumps(vars(settings)))
+
+    def n_queries(self):
+        return len(self.f['results'].keys())
+
+    def review_state(self):
+        "Get internal review state variables from hdf5 file."
```
Comment (J535D165) [inline asreview/hdf5_logging.py]: docstring

File: asreview/hdf5_logging.py [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+            g.create_group("new_labels")
+
+        g = g['new_labels']
+
+        np_methods = np.array(list(map(np.string_, methods)))
+        _append_to_dataset('idx', idx, g, dtype=np.int)
+        _append_to_dataset('labels', labels, g, dtype=np.int)
+        _append_to_dataset('methods', np_methods, g, dtype='S20')
+
+#         bool_array = np.full(self.n_labels(), False, dtype=bool)
```
Comment (J535D165) [inline asreview/hdf5_logging.py]: remove

File: asreview/hdf5_logging.py [hunk 13]
```diff
@@ -0,0 +1,259 @@
+import json
+from datetime import datetime
+from pathlib import Path
+
+import h5py
+import numpy as np
+import itertools
```
Comment (J535D165) [inline asreview/hdf5_logging.py]: import order

File: asreview/hdf5_logging.py [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+
+    def add_settings(self, settings):
+        self.settings = settings
+        self.f.attrs.pop('settings', None)
+        self.f.attrs['settings'] = np.string_(json.dumps(vars(settings)))
+
+    def n_queries(self):
+        return len(self.f['results'].keys())
+
+    def review_state(self):
```
Comment (J535D165) [inline asreview/hdf5_logging.py]: I'm not sure what this is doing. A docstring would be nice. Is this a private method?

File: asreview/json_logging.py [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+
+import asreview
+from asreview.settings import ASReviewSettings
+from asreview.analysis.statistics import _get_labeled_order
+from asreview.analysis.statistics import _get_last_proba_order
+
+
+class Dict_Logger(object):
+    """Class for logging the Systematic Review with no storage."""
+    def __init__(self, *_, **__):
```
Comment (J535D165) [inline asreview/json_logging.py]: A general remark. Think about the structure of the Class. This syntax is a bit odd. 

```
class A(object)
    version = "1.0"
    read_only = True
    def __init__(self):
        pass
```
or 
```
class A(object)
    def __init__(self, version="1.0", read_only=True):
         self.version = version
         self.read_only = read_only
```

Or a combination of both

File: asreview/logging.py [hunk 16]
```diff
@@ -1,230 +1,35 @@
...
-from collections import OrderedDict
-from datetime import datetime
-from pathlib import Path
+import os
 
-import asreview
-import numpy as np
-from asreview.settings import ASReviewSettings
 
+def get_logger_class(fp):
```
Comment (J535D165) [inline asreview/logging.py]: `fp=None`

File: asreview/balance_strategies/triple_balance.py [hunk 17]
```diff
@@ -15,26 +15,36 @@ class TripleBalanceTD(BaseTrainData):
...
         super(TripleBalanceTD, self).__init__(balance_kwargs)
-        self.balance_kwargs['pref_epochs'] = int(fit_kwargs.get('epochs', 1))
-        self.balance_kwargs['fit_kwargs'] = fit_kwargs
         self.balance_kwargs['query_kwargs'] = query_kwargs
-
-    def func_kwargs(self):
-        return triple_balance, self.balance_kwargs
+        self.function = triple_balance
 
     def default_kwargs(self):
```
Comment (J535D165) [inline asreview/balance_strategies/triple_balance.py]: staticmethod

File: asreview/hdf5_logging.py (hunk 3 above)
Comment (qubixes) [reply asreview/hdf5_logging.py]: From what I found, I can put it in the base class, and the HDF5 class will inherit its docstring, right?

File: asreview/hdf5_logging.py (hunk 1 above)
Comment (qubixes) [reply asreview/hdf5_logging.py]: I thought HFD5Logger was a bit ugly, but that's okay I suppose.

File: asreview/hdf5_logging.py (hunk 14 above)
Comment (qubixes) [reply asreview/hdf5_logging.py]: No it is used by the review class to return to a proper state from a log file.

File: asreview/logging.py (hunk 16 above)
Comment (qubixes) [reply asreview/logging.py]: Are you sure about this? To me it doesn't make much sense for the DictLogger to be a default choice, or at least I feel like that is not really appropriate in this case.

File: asreview/json_logging.py (hunk 7 above)
Comment (qubixes) [reply asreview/json_logging.py]: I don't think so, it can be useful to have as a logger that doesn't save anything.

File: asreview/json_logging.py (hunk 9 above)
Comment (qubixes) [reply asreview/json_logging.py]: Not yet I think? If one adds options to the HDF5 one that are not applicable, then it would start to break.

File: asreview/logging/utils.py [hunk 18]
```diff
@@ -3,18 +3,18 @@
 
 def get_logger_class(fp):
     "Get logging class from file extension."
```
Comment (J535D165) [inline asreview/logging/utils.py]: docstring

File: asreview/logging/utils.py [hunk 19]
```diff
@@ -3,18 +3,18 @@
...
-    from asreview.hdf5_logging import HDF5_Logger
-    from asreview.json_logging import JSON_Logger
-    from asreview.json_logging import Dict_Logger
+    from asreview.logging.hdf5 import HDF5Logger
```
Comment (J535D165) [inline asreview/logging/utils.py]: Please move them out of the function (better performance).

File: asreview/logging/utils.py [hunk 20]
```diff
@@ -3,18 +3,18 @@
...
+    from asreview.logging.dict import DictLogger
 
     if fp is None:
-        return Dict_Logger
+        return DictLogger
 
     log_ext = os.path.splitext(fp)[1]
     if log_ext in ['.h5', '.hdf5', '.he5']:
-        Logger = HDF5_Logger
+        Logger = HDF5Logger
```
Comment (J535D165) [inline asreview/logging/utils.py]: `Logger` should be lowercase (and this is not the logger defined below isn't it?).

File: asreview/logging/utils.py [hunk 21]
```diff
@@ -3,18 +3,18 @@
 
 def get_logger_class(fp):
```
Comment (J535D165) [inline asreview/logging/utils.py]: I think we can drop the class `Logger`. The functionality can be combined with this function. 

```
logger = get_logger_class(fp).from_file(fp)
```

This might require some work because of the class attribute `Logger.logger`. But this makes the entire API consistent.

Comment (qubixes) [reply asreview/logging/utils.py]: Well, except it wouldn't work properly as a context manager or would it? Maybe it makes sense to make get_logger_class private?

Comment (qubixes) [reply asreview/logging/utils.py]: It does look kind of ugly with two fp's, no?

Comment (J535D165) [reply asreview/logging/utils.py]: I had a more detailed look at what is going on. Hopefully, I can clarify it a bit. 

I think there should be a function `get_logger(fp)` that initializes `_get_logger_class(fp)`. This function outputs an initiated object that inherits from `BaseLogger`. The class `Logger` can get removed. Context managing is done in `BaseLogger`. 

Throughout the code library, we can replace  

``` python
with Logger.from_file(log_file) as logger:
     if not logger.is_empty():
```
 with
``` python
with get_logger(log_file) as logger:
     if not logger.is_empty():
```

This syntax is in line with other Python libraries. Please have a look at `get_logger` in tensorflow and `open` in base. 

The last one is a good example of the prefered syntax. One can call `with open(fp) as f: `. This returns an `_io.TextIOWrapper` object named `f`. What we do: We call something like `_io.TextIOWrapper.read(fp)` instead of using our util `get_logger`. 

And while we're at it, think about renaming `get_logger()` into `open_logger()`. Completely in line with `open()`

File: asreview/logging/base.py [hunk 22]
```diff
@@ -0,0 +1,225 @@
...
+        self.restore(log_fp)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *_, **__):
+        self.close()
+
+    def __str__(self):
+        print(self.to_dict())
```
Comment (J535D165) [inline asreview/logging/base.py]: This should be a return statement.

File: asreview/analysis/readers.py [hunk 23]
```diff
@@ -0,0 +1,21 @@
+import os
+
+from asreview.logging.utils import Logger
+
+
+def loggers_from_dir(data_dir, prefix="result"):
```
Comment (J535D165) [inline asreview/analysis/readers.py]: Move this function to `logging.utils`. Pretty useful function. Please extend docstring and sphinx reference.

File: asreview/analysis/analysis.py [hunk 24]
```diff
@@ -0,0 +1,307 @@
...
+            y_coef = 1.0
+
+        for i in range(len(norm_yr)):
+            if norm_yr[i] >= val - 1e-6:
+                return (norm_yr[i] - norm_xr[i],
+                        (x_return[i], x_return[i]),
+                        (x_return[i]*y_coef, y_result[i]))
+        return (None, None, None)
+
+    def RRF(self, val=10, x_format="percentage", **kwargs):
```
Comment (J535D165) [inline asreview/analysis/analysis.py]: Lowercase. Can you make those names more explicit?

File: setup.py [hunk 25]
```diff
@@ -42,12 +42,16 @@
...
+        'dill',
+        'PyInquirer',
+        'fuzzywuzzy',
+        'h5py',
     ],
 
     extras_require={
         'dev': ['check-manifest'],
         'test': ['coverage'],
+        'faster Levenshtein': ['python-Levenshtein'],
```
Comment (J535D165) [inline setup.py]: 'faster Levenshtein' -> 'performance'?

**Comment by J535D165** [comment]
The gif is nice. Please add the config file and the raw data to the folder as well. @GerbrichFerdinands can use those to make the other gifs :)

**Comment by J535D165** [comment]
And please rename the gif into something less fancy. 

I think we need gif like `asreview_select_papers.gif`, `asreview_installation.gif`, `asreview_classification.gif`, `asreview_export.gif`. But let's discuss this with @GerbrichFerdinands.

**Comment by J535D165** [comment]
Btw, I think we are close to merging.

File: asreview/logging/utils.py (hunk 19 above)
Comment (qubixes) [reply asreview/logging/utils.py]: I can't; I get import errors when I do that.

File: asreview/analysis/analysis.py (hunk 24 above)
Comment (qubixes) [reply asreview/analysis/analysis.py]: I updated the docstring of the rrf, I think using the abbreviations is okay, since scikit-learn does it in a similar way?

**Comment by qubixes** [comment]
@J535D165 Thanks for the in depth review!

I resolved the Issues that were changed to remove some clutter from the review.

I was thinking that the gif would be a placeholder one, but I made a cleaner version with the config file so that it is at least somewhat presentable. I also updated the 10 minutes to reflect the new version.

Let me know when you're ready to merge, then I'll squash the review commits into one.

File: asreview/logging/utils.py (hunk 21 above)
Comment (qubixes) [reply asreview/logging/utils.py]: I'm still not exactly sure what you meant, but check the new implementation, I think that should work fine, right?

**Comment by J535D165** [comment]
Great work @qubixes!!
